### PR TITLE
(feat) Destination Groups: API Endpoint for Destination Group and Output Channels.

### PR DESCRIPTION
**Pull request by mdhaman**

### CODE REVIEW

File: server/apps/publish/destination_groups.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                message='Destination Group is referenced by items.',
+                payload={'item': 1})
+
+        dest_groups = self.get(req=None, lookup={'destination_groups.group': doc.get(superdesk.config.ID_FIELD)})
+        if dest_groups and dest_groups.count() > 0:
+            raise SuperdeskApiError.preconditionFailedError(
+                message='Destination Group is referenced by another destination group.',
+                payload={'destination_group': 1})
+
+    def _is_self_referenced(self, dest_group_id, dest_groups):
```
Comment (sivakuna-aap) [inline server/apps/publish/destination_groups.py]: Private methods inside a class start with `__`

File: server/apps/publish/destination_groups.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    datasource = {'default_sort': [('name', -1)]}
+    privileges = {'POST': 'destination_groups', 'DELETE': 'destination_groups', 'PATCH': 'destination_groups'}
+
+
+class DestinationGroupsService(BaseService):
+
+    def on_update(self, updates, original):
+        dest_groups = updates.get('destination_groups', [])
+        if dest_groups:
+            if self._is_self_referenced(original[superdesk.config.ID_FIELD], dest_groups):
```
Comment (sivakuna-aap) [inline server/apps/publish/destination_groups.py]: I think this check is needed for create also.

File: server/features/steps/steps.py [hunk 3]
```diff
@@ -342,6 +342,21 @@ def store_placeholder(context, url):
             setattr(context, get_resource_name(url), item)
 
 
+@when('we post to "{url}" with "{tag}" and success')
+def step_impl_when_post_url_with_tag(context, url, tag):
+    with context.app.mail.record_messages() as outbox:
+        data = apply_placeholders(context, context.text)
+        url = apply_placeholders(context, url)
+        print(url)
```
Comment (sivakuna-aap) [inline server/features/steps/steps.py]: Remove `print`. Also I think this is repetitive code.

**Comment by sivakuna-aap** [comment]
:+1:

File: server/features/destination_groups.feature [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    """
+    When we patch "/destination_groups/#destgroup1#"
+    """
+    {
+      "name":"Group 1 testing",
+      "destination_groups": [{"group":"#destgroup2#"}]
+    }
+    """
+    Then we get error 400
```
Comment (amagdas) [inline server/features/destination_groups.feature]: I think we should check the returned error message so it's not some generic message.

File: server/apps/publish/destination_groups.py (hunk 1 above)
Comment (mdhaman) [inline server/apps/publish/destination_groups.py]: private by implementation. there is nothing private in python. we have _ and __ littered everywhere.
But will make this change.

File: server/apps/publish/destination_groups.py (hunk 2 above)
Comment (mdhaman) [inline server/apps/publish/destination_groups.py]: missed that

File: server/apps/publish/destination_groups.py (hunk 1 above)
Comment (sivakuna-aap) [inline server/apps/publish/destination_groups.py]: Agreed... :+1:

File: server/features/destination_groups.feature (hunk 4 above)
Comment (mdhaman) [inline server/features/destination_groups.feature]: @amagdas  done that.

**Comment by amagdas** [comment]
:+1:

**Pull request merged**